kicad: cleanup, fix and update

evils opened the pull request:
###### Motivation for this change
fixes #72248
fixes #49090

includes / closes #72813
affected by #73145, #39493

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
  * +101466360 (whatever units those are...) (comparing old `kicad-with-packages3d` to new `kicad`)
  * +105110176 (old `kicad` vs new `kicad-small`)
  * +5342432736 (old `kicad` vs new `kicad`, now includes the 3D packages)
  * +9955620808 (fixed `kicad-unstable`)
    * of which 3517424376 because `debug = true;`
- [x] Ensured that relevant documentation is up to date
  * i couldn't find any...
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

* start from #72813

* separate the building in base.nix
  * separate derivation, reused for kicad-small
  * source switched to gitlab as development has moved there

* separate the libraries' definitions in libraries.nix

* separate derivations, reused for kicad-small
  * i18n isn't a library but here, gets linked into each base derivation
    * currently the only one moved to gitlab

* a bunch of wrapping in default.nix
  * moved all the revs, hashes and dependency specifics here
  * accept a pname to select which package to build [ kicad | kicad-small | kicad-unstable ]
  * pass the library paths as env vars in the wrapper

* make kicad have the 3D packages by default
  * add alias for kicad-with-packages3d
  * and remove that from all-packages.nix

* add kicad-small to preserve the original kicad behaviour of not including the 3D packages (5gb)

* fix kicad-unstable
  * use wxGTK31
    * brings high DPI support?
    * the python shell in pcbnew fails to open...


###### Status

- [ ] perfect default.nix
  - [ ] kicad from nix on debian
    - [ ] the 3D viewer doesn't work, missing the viewport
      - could be because of my old debian machine
      - probably unrelated, but the only remaining output is fontconfig errors
    - [ ] python error `no module named 'math'` goes away in a `nix-shell --pure`, how replicate this behaviour in an installation?
  - [ ] figure out what meta.platforms to set (kicad is cross-platform)
  - [x] aarch64 builds by using opencascade-occt
      - [x] functionality tested on x86_64
      - [x] functionality tested on aarch64
  - [ ] file chooser not getting the icon for *.pro
    - i seem to recall it worked at some point, but can no longer find it
  - [ ] 5.1.x fails 1 test (eeschema)
    - probably not nix's fault, kicad's master branch tests right

  - kicad-small (without 3D models), should it be?

  - unstable
    - [x] needs a way to override version, and rev and sha256 for each of the separated libraries
      - passing in a `pname` and using that to switch the sources and some other stuff

###### nice-to-haves

- [x] separate the built kicad from the wrapping (to make wrapper debugging quicker)
  - i18n can't be fully separated
  - [x] move all the all-packages.nix stuff to the wrapper (default.nix)

###### reminders to self, before undrafting, do:

- switch all sources to gitlab if they've completed the move by then
- [x] summarize the changes
- the fix for #72248 is temporary, check [the upstream MR](https://gitlab.com/kicad/code/kicad/merge_requests/11)
- [x] condense commits, fully history available [here](https://github.com/Evils-Devils/nixpkgs/tree/kicad-unstable-full)
  - [x] add myself to maintainers list

###### Notify maintainers

cc @berce 
and @Kiwi who was added (on their request) as a maintainer

evils committed:
maintainers: add evils

evils committed:
kicad-unstable: wrap

use wxGTK31
  brings hiDPI support
  no python shell...

matthuszagh committed:
kicad: python2 -> python3

evils committed:
kicad: to hell with version specific crap

matthuszagh committed:
kicad: add python six package as dep

This is needed for python scripting support.

evils committed:
kicad: switch to best try from unstable and 5.1.5

make unstable use kicad-libraries
  still using a link in $out..., not sure that's a bad thing
  this allows setting that path in makeWrapperArgs
    can't use $out there

kicad-with-packages3d -> kicad and kicad-small

default to OCCT, OCE is outdated
  enforce OCCT on aarch64, where OCE is broken
  withOCE flag allows using OCE on non-aarch64

gebner commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -1,11 +1,15 @@
+, fetchpatch
+
 , oceSupport ? true, opencascade
+, withOCCT ? false, opencascade-occt
```
OCCT should probably be the default (instead of OCE) unless there are good reasons for it, since OCCT is the newer version.

Evils-Devils replied:
OCE does seem outdated...
i was going with OCE because this package used it before i touched it, and KiCad uses it by default
the only reason i added OCCT is because OCE fails to build on aarch64
going by [this](https://forum.freecadweb.org/viewtopic.php?t=12920#p103483), i'm probably going to switch it all over to OCCT, unless i hear from their IRC

gebner commented:
Does this PR address the issue with kicad not finding the new libraries after updating?

I.e., right now `kicad` stores the current library path in its configuration:
```
$ grep nix/store -r ~/.config/kicad/
/home/gebner/.config/kicad/kicad_common:KICAD_SYMBOL_DIR=/nix/store/9rjjsly6vf5qawihvgvpf1vcx8dg899k-kicad-5.1.2/share/kicad/library
/home/gebner/.config/kicad/kicad_common:KICAD_TEMPLATE_DIR=/nix/store/9rjjsly6vf5qawihvgvpf1vcx8dg899k-kicad-5.1.2/share/kicad/template
/home/gebner/.config/kicad/kicad_common:KISYS3DMOD=/nix/store/9rjjsly6vf5qawihvgvpf1vcx8dg899k-kicad-5.1.2/share/kicad/modules/packages3d/
/home/gebner/.config/kicad/kicad_common:KISYSMOD=/nix/store/9rjjsly6vf5qawihvgvpf1vcx8dg899k-kicad-5.1.2/share/kicad/modules
```

Even if you update kicad, it still looks for footprints, etc. in the old path.

Evils-Devils commented:
with this PR i add those paths to the wrapper
from a quick test,
it looks like it uses those in favour of the .config ones
it even looks like it doesn't save them, which is probably preferred in this case

evils committed:
kicad: use separated libraries

switch source to gitlab as that's the new upstream source

use wrapper variables for everything but i18n
  add sym and fp templates to template path

update meta to reflect kicad's own language
  set license to AGPLv3, according to the source's LICENSE.README

reduce diff between default and unstable in preparation of merging

on debian gets "no module named 'math'"
  pcbnew 3d viewer is unusable (at least on my T410)

evils committed:
kicad: unite unstable.nix with default.nix

use latest libraries for unstable
move all revs and hashes to default.nix

thanks TQ for getting me through this blockage

evils committed:
kicad: split wrapping and building

evils committed:
kicad: update & cleanup

minor simplification, fix base version

split i18n and get from gitlab

correct wrapper pythonpath, cleanup build output

update & fake git describe

correct base.nix to fit contributing.md

gebner commented:
Works for me.  I can confirm that the issue with the stale paths is resolved now.  @Evils-Devils Is this ready to merge?  (I think the remaining todos should be done in follow-up PRs.)

Evils-Devils commented:
@gebner as this is a rather large overhaul, i wouldn't mind a bit more confirmation that i didn't violate some convention (i can't find any other package using the approach i ended up with).

I do believe it's ready to merge (though i'm about to force push the latest unstable hashes, after i confirm it they work)

As i don't know how to fix the remaining things i'd like done, should i open an issue for those?

gebner commented:
This PR has now been open for a few months.  If somebody had a strong opinion, they would have commented by now.  Since this fixes a lot of issues with kicad, I'm going to merge it.

Feel free to open issues for the remaining todo items if you want.

gebner commented:
> though i'm about to force push the latest unstable hashes, after i confirm it they work

Sorry, completely missed this part.  Please make a new PR for the updated unstable hashes.
